active/0006-extensibility-support-for-deployment-stacks.md

```diff
@@ -0,0 +1,711 @@
+the configuration portion of the extension details will not be returned, as it will likely contain secrets. In order to
+enforce this rule, if the deployment is made from the stacks service, then a boolean property persisted on the
+deployment entity or detail in the job metadata will indicate so to drive how extension configurations are persisted on
+the entity and returned from the GET API.
+
+Examples of expression allowance for the `kubeConfig` case:
+
+```bicep
+@secure()
+param secureParam string
+
+param nonSecureParam string
+
+param boolParam bool
+
+resource kv 'Microsoft.KeyVaults/keyVault@...' = {
+  // ...
+}
+
+resource aksCluster '...' = {
+  // ...
+}
+
+module extResources 'extResources.bicep' = {
+  name: 'extResourcesDeploymentName'
+  extensionConfigs: {
+    k8s: {
+     // NOTE: ❌'s and ❔'s only apply to stacks based deployments. Regular deployments should be able to use any 
+     // valid Bicep expression.
+     ✅ kubeConfig: aksCluster.listClusterAdminCredential().kubeconfig[0].value
+     ✅ kubeConfig: kv.getSecret(...)
+     ❌ kubeConfig: loadFileContent('kubeconfig.txt')
+     ❌ kubeConfig: secureParam
+     ❌ kubeConfig: nonSecureParam
+     ❔ kubeConfig: boolParam ? ... : aksCluster.listClusterAdminCredential().kubeconfig[0].value // allowable, but 
+     blocked by any non-determinism
+     ❔ kubeConfig: userDefinedFunctionCall()
+    }
+  }
+  params: {
+    // ...
+  }
+}
+```
+
+The expressions in question depend on how difficult it would be to detect usages of invalid expressions recursively. The
+deployments engine could evaluate these expressions out to values, but it would need to be able to determine the source
+of any data used to arrive at the final value to ensure secrets are not leaked. It may be best to only allow direct ARM
+resource access for secret config values initially and then expand allowance in future enhancements.
+
+#### Stacks mode for Bicep files
+
+Because the stack service will enable additional validation and features in the deployment service, Bicep template
+authors need to see diagnostics that flag disallowed expressions ahead of time. With the current Bicep file
+architecture, this cannot be determined automatically. Bicep files can be used with or without stacks and modules bind
+together different Bicep files to create a deployment hierarchy. To enable the additional diagnostics, they will need to
+be explicitly enabled by the template author. Here is the design for how to enable this in a Bicep file via the Bicep
+linter:
+
+```json5
+{
+  "rules": {
+    "stacks-compatibility": {
+      "level": "warning",
+      // The default level across all files. Up to user to configure as an error or to turn off.
+      // Alternative to file overrides is to stick with in-file/line overrides.
+      "fileOverrides": {
+        "*.stack.bicep": "error",
+        "<glob>": "error"
+      }
```

Comment:
Bicep discussions 10/14: 
- Make it consistent with existing naming schemes.
- Break into smaller rules for maintainability.
- Default to informational.
- Hold off on file overrides for now. Collect feedback once rule is released.